### ISSUE TITLE
Make incremental compile efficient in the presence of Annotation Processors

### ISSUE THREAD

**Issue by eriwen** [gradle/gradle#1320]
Migrated from [GRADLE-3259](https://issues.gradle.org/browse/GRADLE-3259) which had 19 votes

Compile avoidance was [introduced in Gradle 3.4](https://docs.gradle.org/3.4/release-notes.html#compile-avoidance-in-the-presence-of-annotation-processors) (<= Users please read this). One can improve performance somewhat by separate annotation processors from the compile classpath. 

### Expected Behavior
Incremental Java compilation should avoid full rebuild when annotation processors are present. 

### Current Behavior
Currently we fall back to a full recompile when annotation processors are used. The reason for this is that we don't know which source files they need as an input. The APT API allows processor to access other source files then the one containing the annotation being processed.

### Context
We could instrument the compiler to find out which files the annotation processors are actually reading and use this information to run them incrementally as well.

In cases where an annotation processor is taking the whole source tree as an input, we could warn the user (e.g. in build scans) that this is causing slower builds. Many developers might not even be aware of the trade-off between "saving time writing code" and "saving time running the build" they have to make.

### Design Spec

Since this is a bigger feature, you can download the [design spec](https://github.com/gradle/gradle/files/1659294/Incremental.Annotation.Processing.-.Public.Copy.pdf) as a PDF.

**Comment by stephanenicolas** [comment]
Great news ! 

Instead of instrumenting the compiler, it would be possible also to apply the same discovery mechanism for annotation processors as APT  and ask all processors which annotations they process, and find all files containing these annotations. 

It could also be interesting to let the javac maintainers know about this work when it's done as it should lead to an enhancement of the APT APIs / JSR. They could also probably help to instrument the compiler if that's the way to go.

**Comment by adammurdoch** [comment]
Yes, this is what "instrumenting the compiler" means.

**Comment by stephanenicolas** [comment]
I thought it was about hacking it via reflection or something like that.
Thanks for the precision.

We would actually be glad to participate if there is some space for
collaborative work.

2017-02-05 12:31 GMT-08:00 Adam Murdoch <notifications@github.com>:

**Comment by dongshengbc** [comment]
The link to "introduced in Gradle 3.4" does not work?

**Comment by eriwen** [comment]
@dongshengbc Fixed it. Sorry I had linked to the release candidate before.

**Comment by stephanenicolas** [comment]
@eriwen , we are really very motivated to get this solved. Can we help in anyway ?

**Comment by eriwen** [comment]
@stephanenicolas Thanks for volunteering. I think there are 4 potential ways to help:
 * Technical design or implementation of specific items (delegating to @oehme on this)
 * Provide really great, generic test cases so we can more easily determine effectiveness of work
 * Try out anything currently under development, providing build/technology details (I'm not keen on the plan, but Stefan may be)
 * Consider sponsoring some of the features through Groupon, so that we can afford to offload non-critical work to others (or new engineers) so our core engineers can focus on critical work like this. Let me know if Groupon is interested in this and I'll put you in touch with the right people.

@oehme leads the team that would be responsible for addressing this. Anything technically they (Groupon team) can take on here?

**Comment by stephanenicolas** [comment]
I will come back to you on this. If you could have a more detailed list of areas with which we can help, that would be great.

**Comment by stephanenicolas** [comment]
@oehme @eriwen is there a way we can get more directly in touch with you and work together on this. We are switching to gradle 4 and the new android plugin and this issue is really our last bottle neck to get decent build times. If you wanna write an email to me, my personal email address is pretty easy to find on the web..

**Comment by meierjan** [comment]
Any progress? I agree with/can just quote @stephanenicolas from the old issue tracker:

> This is probably the most fundamental issue in Android builds that impacts developers productivity in our team, and actually probably the main motivation for us to soon consider moving to another build system. That's quite a pity !
> FYI : https://code.google.com/p/android/issues/detail?id=200043

I wonder why this is not having a much higher priority because I think this is an issue for many (many) developer - or am I forgetting something?

**Comment by oehme** [comment]
We are working together with the Google and Groupon teams, but the work is still at an [early stage](https://github.com/gradle/incap). We plan to have a first milestone for simple processors like AutoValue and Butterknife late October.

> I wonder why this is not having a much higher priority

There are only so many things we can tackle at once. I understand that this is probably the biggest issue for you. But I hope you understand that there are other big issues for other users.

**Comment by jaredsburrows** [comment]
@oehme @eriwen What is the status of this? Adding support in Gradle for Android builds would greatly reduce build times since a lot of projects use libraries such as Glide and Dagger 2.

**Comment by lukyer** [comment]
If anyone interested in how to achieve full gradle power (e.g. incremental builds) with annotation processors, see my workaround solution here https://stackoverflow.com/questions/47289457/how-to-configure-gradles-incremental-build-with-annotation-processor/47392788#47392788

**Comment by oehme** [comment]
@jaredsburrows see #3434 for the current status. Dagger 2 will be complicated though, because even if just one class changed, it would still need to do some heavy analysis to make sure all injection points are still valid.

**Comment by oehme** [comment]
I've broken this epic down into a few initial stories. You can take a look at the [design spec](https://github.com/gradle/gradle/files/1659294/Incremental.Annotation.Processing.-.Public.Copy.pdf) to get a better idea of what use cases we plan to support and what is out of scope (for now).

**Comment by fab1an** [comment]
A good solution. Only one question:

> When a processor does not opt-in (no properties file provided), we do full recompiles as
before

Why not fail the compilation until incremental is switched off here as well?

**Comment by oehme** [comment]
For backwards compatibility reasons and because we want to remove the incremental flag altogether in the mid term, making incremental the default and only mode.

**Comment by pbi-qfs** [comment]
The approach sounds reasonable and well thought out. I would agree with @fab1an to fail fast instead of silent full recompile, that would be more consistent.

In addition, it would be handy if the build composer would have an option to provide the processor type information in the build file, if none is provided by the processor author, e.g. in the configurations section.

**Comment by oehme** [comment]
> In addition, it would be handy if the build composer would have an option to provide the processor type information in the build file, if none is provided by the processor author, e.g. in the configurations section.

I thought about this too, but I'm not sure in how many cases it would work, since the processor also needs to use the originatingElements API correctly.

**Comment by pbi-qfs** [comment]
> I thought about this too, but I'm not sure in how many cases it would work, since the processor also needs to use the originatingElements API correctly.

In the most simplest case, a processor (like Lombok) only modifies the one class defined by the corresponding java, and might generate additional inner classes, which can easily be associated by the $-name pattern. Taking all $-classes into account is most probably overkill, but it is a good default if the processor doesn't know about the originatingElements API, yet - it's better than a full recompile.

**Comment by oehme** [comment]
APT does not allow processors to change classes in-place. Lombok is operating outside the APT specification and uses internal APIs. I would not consider it a "simple case". There is also no way to generated inner classes. You might use a dollar sign as a separator, but that won't be considered an inner class by the compiler (and it won't have access to the internals of the outer class).

For any generated file we will require originating elements to be provided, otherwise processing will fail. That's why I think the number of processors that use this API correctly and don't have an opt-in declaration will likely be small.

**Comment by pbi-qfs** [comment]
> For any generated file we will require originating elements to be provided, otherwise processing will fail. That's why I think the number of processors that use this API correctly and don't have an opt-in declaration will likely be small.

I totally agree, therefore I'm looking for a way to "approximate" a solution better than a full recompile for processors which do not know about the Gradle API, yet. Such an approximation would be the association of dollar-classes with their "main" class.

**Comment by oehme** [comment]
Most processors don't use a dollar sign (some use underscores, some just use camel case) and I really don't want to put such assumptions in, as correctness is the most important concern for Gradle. Let's first see if convincing processor authors to opt in is a problem at all. If it isn't then we've saved a bunch of time. If it is, we can start thinking about alternatives.

**Comment by Alexander--** [comment]
What is the current state of this issue?

Looks like https://github.com/gradle/incap and https://github.com/gstevey/gradle have been inactive for months with some changes being mainstreamed in 2bc39580e49c55b6c97c04ef01c4a224b62ede66 - db119def446f92276399172e803f1cf0156c353d. Is there still place for discussion? Will that be the final implementation?

**Comment by oehme** [comment]
@Alexander-- did you read my [update](https://github.com/gradle/gradle/issues/1320#issuecomment-360075109) and design spec?

**Comment by Alexander--** [comment]
@oehme Yes, I have read [this document](https://github.com/gradle/gradle/issues/1320#issuecomment-360075109) (if that's the one you are talking about), but it has only high-level overview of design. Many details are omitted, some potential issues are clearly downplayed.

For example, that document does not describe, how processors will to communicate with Gradle and what changes need to be made to existing processors to "support" incremental compilation.

In additions, the spec mentions making a JSR to improve `javax.lang.model` API as one of alternatives to the currently chosen approach. It would be great to hear more about that, — what kind of limitations in javac and Mirror API are stopping Gradle from doing it's part? The Mirror API has supported incremental compilation for many years and the chosen approach already takes advantage of that support. For me it sounds like that part of javac is already *perfect*.

**Comment by oehme** [comment]
Processors don't communicate with Gradle directly, they just use the existing Filer API as mentioned in the document.

What potential issues are omitted/downplayed?

**Comment by Alexander--** [comment]
@oehme 

> Processors don't communicate with Gradle directly, they just use the existing Filer API as mentioned in the document.

Yes, that part is great.

> What potential issues are omitted/downplayed?

"what changes need to be made to existing processors to support incremental compilation."

**Comment by oehme** [comment]
I'm not sure what else you are looking for, but I might be misunderstanding your question. 

The two requirements are listed in the doc: Provide originating elements when you call the Filer API (which many processors do already for better diagnostics) and opt into incremental processing with a simple CSV file (the format will be published in the Gradle documentation). There's nothing else to do as a processor author.

**Comment by Alexander--** [comment]
@oehme 

Are you sure? Please correct me if I'm wrong, but from looking at at https://github.com/gradle/incap I got an impression, that each processor will have to extenda Gradle-specific base class and embed a chunk of dependency tracking logic (by depending on Gradle-provided library). Is this still the case?

**Comment by oehme** [comment]
@Alexander-- No, that project will be deleted. Gradle will extract the information it needs transparently. 

Coming back to the JSR idea: The compiler API doesn't make it easy to do incremental processing, as there is no way to distinguish a single-origin and multiple-origin processor. I.e. if the user gives us one originating element for a generated file, this could mean that this is a 1:1 mapping from source to target. But it could also be many:1 and the "many" was 1 at this point, but could grow later. These cases need to be treated differently. Of course it is doable, if you run the compiler once, cross-reference the previous originating elements against the current ones and then resping the compiler when you see that this generated file had other sources before. But that means taking even more responsibility in Gradle, which should really be handled by javac instead.

Taking this a bit further - Ideally javac should take over the whole incremental state handling and Gradle (or IntelliJ etc.) should only have to provide the files that physically changed as an argument. That's what I'd consider true support for incremental compilation.

**Comment by Alexander--** [comment]
@oehme 

> No, that project will be deleted, thanks for the reminder! Gradle will extract the information it needs transparently.

Thanks, that's great to hear.

> Coming back to the JSR idea: The compiler API doesn't make it easy to do incremental processing, as there is no way to distinguish a single-origin and multiple-origin processor
> …
> **But it could also be many:1 and the "many" was 1 at this point, but could grow later.**

I had trouble understanding that distinction, and this line of yours finally did get it across. But doesn't calling non-conforming processors "many:1" hide a bigger problem? It sounds like you are trying to simultaneously pursue two massive goals:

1. Making a primary compilation incremental in presence of processors
2. Making processors themselves incremental (or at least reasonably efficient in absence of changes)

Could you please address those issues separately? Also names like "1:1" and "many:1" are very misleading and have nothing to do with processors themselves being compatible with incremental compilation. Those knobs should not be tied together.

**Comment by oehme** [comment]
I think you misunderstood my point above. I'll try to explain with an example:

Let's assume we have one processor that takes classes annotated with `@Helper` and generates a `xxxHelper` class for each of them. This is what we call "single origin". Now let's say I already have a class `@Helper class A {}`. I now add `@Helper class B {}` to my sources. In this case, `B` can be compiled on its own and we don't need to add `A` to the compile run, as the generated sources for `A` and `B` have no connection.

Now let's look at a processor that takes **all** classes annotated with `@Service` and generates a **single** `ServiceRegistry` with one method for each service. Again, I already have a class `@Service class A {}`. But now when I add `@Service class B {}`, I can't compile just `B` alone. That would result in a `ServiceRegistry` that only knows about `B`. To be correct, we also need to recompile `A`.

From the `Filer` APIs perspective, both cases initially look the same - A single originating element. But the behavior needs to be different when a second element is added. That's why we require processor authors to categorize their processors into "single origin" and "multiple origin".

**Comment by Alexander--** [comment]
@oehme

> I think you misunderstood my point above.

Did I? You are clearly conflating overall incremental support with each processor itself having an optimal incremental path.

There are popular processors, like [this one](https://github.com/pellaton/spring-configuration-validation-processor), that validate annotated classes for to matching convention without generating new classes. Those don't currently fit into your model — there is no generated classes, thus no way to specify explicit originating elements. But such processors are still compatible with incremental compilation — you just need to treat all elements, annotated with target annotations, as single implicitly specified dependency (which, I suppose, is how you handle "multy:1" processors).

Now let's take a step further — a processor, that targets `"*"` to generate `RegistryOfAllClassesInProject`… or even N such classes (e.g. a genuine "many:many" processor). Such processor can also be made compatible with incremental build — provided, that Gradle adds all previously built classes to compilation classpath of subsequent builds. You just need to treat each generated file as if it depends on each file in project. Such incremental compilation might not be optimal for *that processor* — it will have to visit all files in project on each build, and all generated files have to be removed on each build. But it is still better for javac as whole, — because the compiler itself doesn't need to recompile whole project on each build.

**Comment by oehme** [comment]
> Those don't currently fit into your model — there is no generated classes, thus no way to specify explicit originating elements.

Which shows that the Compiler API is far from perfect. These validating processors will have to decide on a case by case basis if they fit into the constraints of the "single origin" category (i.e. each of their warnings has one originating element).

> Such processor can also be made compatible with incremental build — provided, that Gradle adds all previously built classes to compilation classpath of subsequent builds.

Gradle already does that, but I fail to see how the processor would work incrementally, since the RoundEnvironment would only give the processor the changed sources, not a listing of previously generated classes.

**Comment by Alexander--** [comment]
> Which shows that the Compiler API is far from perfect.

I think, that it is mostly perfect. Access to internal structure  of methods is prohibited. Modifying existing classes is prohibited. Gradle should be able to work out the rest.

> These validating processors will have to decide on a case by case basis if they fit into the constraints of the "single origin" category (i.e. each of their warnings has one originating element).

That doesn't make sense, — warnings don't have originating elements. There is a support for specifying a single causal element per `Messenger` method call, but those aren't same as originating elements, and those elements might not be the only ones, contributing to each warning. Regardless, your spec currently fails to account for those processors: it says, that empty set of originating elements makes Gradle fail incremental build. Clearly, there should be support for other sources of originating elements and/or implicit originating elements.

> I fail to see how the processor would work incrementally, since the RoundEnvironment would only give the processor the changed sources, not a listing of previously generated classes.

You have control over RoundEnvironment, so you should be able to fix it. Or just let it be (for now) and declare supporting that a prerequisite for supporting incremental build by such processors.

Anyway, we are sidetracking from my main point, — "1:1" and "many:1" are poorly chosen terms. They don't accurately describe the problem at hand. For example, a processor might generate exactly one source file per original source file, but determine *which* files to handle by fetching a JSON from internet. Such processor can still support incremental build, but it won't play well with your "1:1" optimization (because it is not idempotent).

**Comment by oehme** [comment]
> your spec currently fails to account for those processors:

It accounts for most of them. If you don't generate any sources, then you need to live with Gradle's default incremental behavior, which is to recompile the files that are affected by a change (a file is affected if it physically changed or if it depends on an affected file). If your processor can live with these semantics, you can declare it as incremental and it'll be just fine. If you create a single warning from multiple source elements (e.g. "You have too many Foos in your code base"), then you're out of luck. There is no way Gradle could know what this processor is trying to do, because the javac API doesn't give it a way to express "These types all need to be revalidated together".

> For example, a processor might generate exactly one source file per original source file, but determine which files to handle by fetching a JSON from internet. Such processor can still support incremental build, but it won't play well with your "1:1" optimization (because it is not idempotent).

That's a strawman argument. A processor could even format your hard drive, sure. That's why we'll never support all of them. The processors that are actually in widespread use fall into the two categories we are planning to support. All other processors will be treated with a full recompile just like before. We can always add support for more categories of processors, but for some there is just not enough information in the javac API.

You repeatedly said the javac API was (almost) perfect, but it can't handle the cases you brought up yourself. Saying that "Gradle should be able to figure it out" is too hand-wavy for me. Please explain *how* Gradle would be able to figure out these usage patterns in a transparent way, given the current javac APIs and we'll certainly consider it.

**Comment by Alexander--** [comment]
> javac API doesn't give it a way to express "These types all need to be revalidated together"

Javac API allows processors to target `"*"`, which means, that they want to run during *all* build rounds (in multi-round build). I believe, that Gradle should provide the same ability in regard to incremental build (along with ability to take advantage of it).

> If it can change its mind which files it'll touch on every invocation then you alway have to do a full compile.

Now you are doing it again. You have to do full run with *that* processor. Not full recompile with javac as whole. Delete *that processor's* files, run *it* over entire project (most of which is already compiled). Don't have javac recompile everything. The full build can still be incremental, only that one processor won't benefit from it (it's own run time won't be reduced).

> The processors that are most widely used…

Citation needed. I'd argue, that most notorious ones don't quite fit into the mold you are defining here (both in respect to error reporting and subtle semantics).

> All other processors will be treated with a full recompile just like before

It is better to have simple, reliable, but a bit less efficient incremental build then nothing at all. IMO, you should make a model acceptable for most processors, then iterate from there.

> Please explain how Gradle would be able to figure out these usage patterns in a transparent way, given the current javac APIs and we'll certainly consider it.

Ok, let me try.

**Comment by Alexander--** [comment]
@oehme 

1. Treat all elements, passed to Messenger and Filer, as *explicit target elements*
   1. Explicit target elements, passed to Filer, are scoped per associated file
   2. Explicit target elements, passed to Messenger, are global for that processor
2. Treat processor's target annotations, annotations, annotated with target annotations, and elements, annotated with those annotations, as *implicit target elements*. Sometimes that can be disastrous (e.g. something like `@Override` or `@Documented`), but whatever.
3. Processors shouldn't assume, that initial RoundEnvironment always contains annotated targets, contains only files under compilation or all modified files. It can contain anything, that needs to be processed according to Gradle's opinion. Processors mustn't opt into incremental build, unless they agree with this.
4. Processors, that rely on getting values of inlineable constant at runtime (by calling Elements#getValue) mustn't opt into incremental build
5. Provided, that a processor opts into incremental build (or user opts it in via build.gradle switch)…
   1. If processor targets `"*"`, assume, that it wants to run during all builds. Don't use originating elements for deletion of generated files, — everything in project (including resource files) is treated as single composite "originating element" in respect to those files. Gradle deletes everything, created by that processor via Filer, before consequential incremental steps.
      1. Promiscuous processors should be idempotent in respect to what elements they report as originating (if any).
      2. Promiscuous processors, that haven't reported any originating elements, aren't eligible for filled-in RoundEnvironment.  They mustn't opt into incremental compilation unless they can deal with it.
      3. Promiscuous processors must report any annotations of interest as their explicit originating elements. Gradle gets elements annotated with those annotations from javac and adds those to initial round's RoundEnvironment during next incremental build.
      4. New originating elements, reported during subsequent incremental builds, cause full rebuild.
   2. If processor targets fixed set of annotations, treat that as opting into the "fast path". If incremental processor does not comply with assumptions of "fast path", it must be changed to target `"*"` to work with incremental compilation (or opt out of incremental build completely)
      1. A "fast path" processor must have fixed set of annotation dependencies
      2. A "fast path" processor must be idempotent in respect to it's inputs
      3. A "fast path" processor must not have sources outside of Gradle's control (no separate configuration files or environment vars, all arguments must be supplied with `-A`)
      4. A "fast path" processor shouldn't inspect elements without reporting them as originating. It mustn't use such elements as base for error reporting or code generation. This means, that "fast path" processors must account for any performed validation when reporting their explicit target elements (for example, by issuing a MANDATORY_WARNING, NOTE or OTHER for each validated element). Processors, that validate unannotated elements without mentioning them via Messenger mustn't opt into fast path.
      5. If "fast path" processor reports empty set of originating elements (for example, because it only does validation and detected no errors, or because it didn't have to generate anything), assume, that all elements, annotated with it's target annotations are it's single composite "originating element".
      6. If "fast path" processor reports explicit originating elements, append implicit ones to those.

**Comment by oehme** [comment]
> It is better to have simple, reliable, but a bit less efficient incremental build then nothing at all. IMO, you should make a model acceptable for most processors, then iterate from there.

The problem is that such a mode is not simple to build and thus not easy to make reliable. We'd have to take complete control of the round environment and other services, provide annotated elements from the classpath etc. All these things are currently untouched by Gradle and left to the compiler.

In general, we don't start with broad solutions that help everyone only a little bit. We start with narrow solutions that help an important subset a lot. Then we build broad solutions later once people tell us about important limitations.

Maybe we are just disagreeing about terminology here, because I see no conflict between these approaches. I'll try to make a full categorization, taking into account your proposal as well:

1. Processors that are okay with Gradle's current incremental compile semantics. All they have to do is provide an originating element whenever they generate a file, so Gradle can clean it up. This is what we currently call "single origin", but I'm open for other names.
2. Much like 1, but with the option to provide multiple originating elements when generating files, so these files are given back to the processor as one chunk. Unfortunately this doesn't work for warnings, since the compiler API does not provide a way to give originating elements to warnings. But it does work for generated files.
3. Reprocessing everything that the processor was interested in, but loading those elements mostly form the classpath instead of reparsing them. This would work for a lot of processors, though less efficient than the two modes above. 
4. Full recompile - The default if you don't make a choice.

I think 3 (your proposal) is a great idea that we should pursue in the future. I don't think it is simple though and thus shouldn't be the first thing we do.

**Comment by Alexander--** [comment]
> I'll try to make a full categorization, taking into account your proposal as well…

The more you reiterate over that classification, the more arbitrary it sounds to me.

For starters, would you kindly explain, what is the practical difference (according to you) between 1st and 2nd types in your list? Please, don't just say "Gradle handles them differently" or "1st is more efficient", — how does Gradle's own behavior differ in respect to those kinds of processors?

You still haven't answered my initial question — what stage is this issue on? As in: are you still working out the correct approach or is there already an implementation, which may soon ship mostly as-is? I am asking because you are talking a lot about deleting generated files, handling of originating elements and dependency groups, but haven't mentioned anything about handling of target annotations and supporting validation-only processors. Those are significant omissions.

If possible, I'd like to help with work on this issue, but it is sounds unclear, if there is still space for that before the current support is released, and what is your vision for next iteration.

**Comment by fab1an** [comment]
Completely wild thought from the side: Wouldn't the "easiest" solution be to provide a virtual filesystem for javac to compile into and track file-changes on that level?

**Comment by Alexander--** [comment]
> Completely wild thought from the side: Wouldn't the "easiest" solution be to provide a virtual filesystem for javac to compile into and track file-changes on that level?

@fab1an 

AFAIK, the current implementation already does something along those lines (by hijacking processor's `Filer`). But processors have other means to interact with files — for example, by querying javac for information about annotated targets (RoundEnvironment#getElementsAnnotatedWith). Such inquires don't necessarily hit javac's filesystem abstraction level, — the compiler may satisfies them by returning cached data. There is no reliable way to associate introspection actions with each individual processors; both false positives and false negatives would appear.

The interface, informing Gradle about processor's dependencies, need to be explicit.

**Comment by oehme** [comment]
> If possible, I'd like to help with work on this issue

@Alexander-- I'm happy to hear that and I think you have good ideas that I'd love to discuss further! But the condescending tone you used so far makes it very hard for me to take the things you say at face value. It makes me feel frustrated. If you could change that, I'd be happy to schedule a hangout to discuss annotation processing with you. This would probably be more efficient than  async communication here.

Assuming the answer is yes, here are some comments on the points you made above:

> what stage is this issue on

We are currently implementing the design as described in the document, parts of it will be in Gradle 4.6 and the rest in Gradle 4.7. This design has been deeply discussed with the Google and Groupon team and several annotation processor authors were involved. Specifically we are targeting the top annotation processors of the Android world first: DataBinding, ButterKnife, Dagger, AutoValue among a few others. From our analysis they mostly fall into category 1 and a few into category 2.

> how does Gradle's own behavior differ in respect to those kinds of processors?

The second category needs to be handled differently because of the issue I explained with the `ServiceRegistry` example earlier. Concretely the first step will be to just recompile these clusters every time, no matter what changed (this is explained in the document, but I'm happy to adjust the wording for the final user guide to make this clearer). This makes these processors much less efficient than category 1. They still need much less infrastructure/responsibility on the Gradle side than your proposed category 3, which is why we are going for this as the first step.

We don't expect to stop here, this is just the first step to unblock the popular processors mentioned above. As we uncover other categories we'll make the support more sophisticated and might end up putting more responsibility on Gradle, possibly implementing your proposal.

> but haven't mentioned anything about handling of target annotations and supporting validation-only processors. 

The targeted annotations don't matter for category 1 and 2, that's why the document makes no restriction on them. Validation-only processors will work if they are fine with the default incremental semantics (i.e. category 1). We can call this out explicitly in the user guide.

**Comment by oehme** [comment]
@fab1an Gradle already knows what files are an input and output for the compiler and whether they changed. But the crucial part is knowing the connection between input and output. This is something you can't tell from looking at file system operations. You can tell it from usage of the `Filer` API though, as long as the processors use it properly. That's why it'll be opt-in for every processor.

**Comment by eriwen** [comment]
To me, the fact that the Android Tools team and other annotation processor maintainers are onboard is good validation that @oehme's specification is the right direction for now, and fortunately we won't have too long before we'll be specifying about how to properly handle more situations where annotation processors are used.

**Comment by simtel12** [comment]
How much of this made it into 4.6?  And will existing annotation processors (Butterknife, AutoValue, Dagger) need to update anything to take advantage of the changes in 4.6?

**Comment by oehme** [comment]
Gradle 4.6 only contained some validation to help early adopters try out their processors. 4.7 will contain the whole feature. Annotation processor authors need to opt in and make sure their processors are using the compiler API correctly, as mentioned in the design document.

**Comment by simtel12** [comment]
@oehme Thanks for the info!

**Comment by oehme** [comment]
Gradle [4.7-20180315192435+0000](http://services.gradle.org/distributions-snapshots/gradle-4.7-20180315192435+0000-bin.zip) contains all the necessary changes and the [documentation](https://docs.gradle.org/nightly/userguide/java_plugin.html#sec:incremental_annotation_processing) of this feature is in the user manual for the nightly.

As you can see the categories have slightly evolved from the original design document and should cover more cases now and be easier to understand.

There are still some optimizations that I want to do, but this is already in a shape for processor authors to start categorizing their processors and trying out the feature.

**Comment by blindpirate** [comment]
Since the majority work has been done, removed `4.7RC1` milestone. Added `4.8` for further optimization.

**Comment by filiphr** [comment]
I am one of the members of the [MapStruct](https://github.com/mapstruct/mapstruct) and I have to say really good job and providing such functionality. I read the design spec and the documentation. It is not quite clear to me which category should I put our processor in. I also don't know that much about gradle and it's incremental compilation as well. Let me summarise what I understood.

In the design spec there is this example:
```java
@Repository
interface CustomerRepository {
    Customer findById(long id);
}
```

And there it is written:

>The generated `CustomerRepositoryImpl` is created because `CustomerRepository` existed. Of
course the contents of `CustomerRepositoryImpl` will also depend on `Customer`, but that
relationship is already tracked, because `CustomerRepository` also depends on `Customer`

And then there is:

>This solves the case of "simple" processors which create one or more generated files from a single
source file

From what I could understand from the documentation is that the processor using `@Repository` is an isolating one. Am I correct? 

I couldn't also understand on which change the Processor would be triggered. Is it only when `CustomerRepository` is changed, or also when `Customer` changes. From the design spec I understand that it would be in both cases. Maybe this needs to be added to the Incremental Annotation Processing part of the documentation. Maybe it is obvious for people that know the Gradle incremental compilation. However, to other annotation processor authors it might not be known.

**Comment by Alexander--** [comment]
@filiphr your processor uses `Filer#createResource`. This is not supported by the current incremental processing infrastructure. Citation from spec:

> Isolating processors have the following limitations:
> …
>    Can’t write resources
> …
> Aggregating processors have the following limitations:
> …
>    Can’t write resources (this may change in the future)

You should do nothing and wait until Gradle updates it's implementation.

**Comment by blindpirate** [comment]
@filiphr 

> From what I could understand from the documentation is that the processor using @Repository is an isolating one. Am I correct?

Yes, you're correct.

>  Maybe this needs to be added to the Incremental Annotation Processing part of the documentation.

Thanks for advice.

**Comment by filiphr** [comment]
@Alexander-- I don't think our users are using that feature that much. And I can also have a look at that part. We are generating that resource is created it is always accompanied with creating another Java class. We are not generating only resources.

@blindpirate thanks for the clarification

**Comment by filiphr** [comment]
Thanks for clearing it up @oehme. To be honest I read only the part about the annotation processor (didn't read the previous chapter). You don't have to repeat, if you just reference the part that it is explained it would be much easier for annotation processor writers 😄.

Yes in our case I would say that it is used in max 1% of the cases. As it requires multiple conditions to be fulfilled and requires actions from the user. So out of the box it can never happen. It's an opt in feature. So if you can do the improvement and not error out it would be great.

Yes we use the `Filer` API to generate the resource. Although, the resource name might change if the user does some changes in the annotation definition. I am going to watch that issue as well.

I just saw that we are not passing the originating elements to the `Filer` API, so we need to do that enhancement on our side.

**Comment by wrotte** [comment]
Is it possible to force gradle to avoid a full rebuild even when annotation processors are present, even if they don't/haven''t implemented this?  For example, in a CI pipeline, I don't want to rebuild modules that have such annotation processors in every stage, but I'm seeing that behavior.

**Comment by oehme** [comment]
@wrotte I'm afraid I don't understand your question. This is only about incremental compilation. It doesn't affect up-to-date checking or caching. Can you elaborate on your setup a bit more?

**Comment by wrotte** [comment]
@oehme Sorry for the late reply; I've been trying to distill a succinct test case and I think I've found that the latest Android plugin solves the issue I'm seeing.  If I can reproduce I'll file a new issue.

**Comment by oehme** [comment]
Good news everyone, the upcoming Gradle 4.8 will contain several big improvements to incremental annotation processing:

- Processors will be able to decide [dynamically](https://github.com/gradle/gradle/issues/4915) if they are incremental or not. This was brought up by the AutoValue team, because they have an extension mechanism and so they first need to check all extensions for incrementality before they can say "Yes, this processor is incremental". This new mechanism will supersede the META-INF registration. 
- Compilation [will no longer fail](https://github.com/gradle/gradle/issues/4960) when a processor does something that we know won't work incrementally. Instead Gradle will simply do a full recompile and inform the user about the problem.
- Unused non-incremental processors [will no longer prevent](https://github.com/gradle/gradle/issues/4961) incremental compilation. This means that incremental and non-incremental processors can be safely packaged together.

**Comment by oehme** [comment]
I've compiled a list of [popular processors](https://github.com/gradle/gradle/issues/5277) and their current state of adoption. Please express your support by giving 👍 on the issue trackers of the ones you care about. Also feel free to suggest others that we should get in touch with.

**Comment by oehme** [comment]
That's not coming from us, but probably from the Android plugin. @ducrohet can you comment on this? It would be great if the incremental compile decision was completely left to Gradle. Is there any adverse condition we are not detecting?

**Comment by ducrohet** [comment]
For some reason our plugin output this when you apply the retrolambda plugin but I'm not sure why. This was done a while back (2 years) and may not be required anymore. I'll investigate.

**Comment by ducrohet** [comment]
It looks like previously this disabled inc java on both android-apt and retrolambda. Probably a remnant of a time when the java compile task didn't automatically disable inc compilation if annotation processor was one.

**Comment by tprochazka** [comment]
I'm definitely not using retrolamba plugin. I tested it with android build plugin 3.1.2. and 3.2.0-alpha15 and it happens with both. I switched butterknige and Android-state to a version which should support incremental compilation.

**Comment by ducrohet** [comment]
yes, sorry, this is triggered with a few more cases than just retrolambda.
The test is: data binding enabled OR annotation processor OR retrolambda.

This still seems a bit strange to me. I'll review this.

**Comment by tprochazka** [comment]
Thanks, so data binding is the reason. Somebody on Google I/O told this year, that one of the news in new building tools is incremental support. So would be great if also java compilation will be part of this, especially now when Gradle is able to do it also with annotation processors.

**Comment by tprochazka** [comment]
I created an issue for "Incremental Java compilation disabled in variant ... as you are using an incompatible plugin" here https://issuetracker.google.com/issues/110061530 to not spam here about the issue which is not related to Gradle itself.

**Comment by oehme** [comment]
I'm closing this issue since the work on the Gradle side is done and lots of annotation processors can now be incremental. Please have a look at the list of [processors](https://github.com/gradle/gradle/issues/5277) we know about and their adoption status. Any kind of support from your side will help. You could

- tell the processor author how much full recompiles are affecting your daily work
- point them to the [documentation](https://docs.gradle.org/current/userguide/java_plugin.html#sec:incremental_annotation_processing) and help them understand the contraints of incremental processing
- send them a PR
- help them with exploratory testing to ensure their processor behaves as expected

I'll be happy to help too, just mention me in a conversation if clarification is needed.

**Comment by oehme** [comment]
They're currently working on making KAPT compatible.

**Comment by oehme** [comment]
I haven't found a public issue, feel free to create one on [Youtrack](https://youtrack.jetbrains.com/issues).

**Comment by TWiStErRob** [comment]
@oehme this one? https://youtrack.jetbrains.com/issue/KT-23880
